src/backends/perf_pt/collect.c

```diff
@@ -283,6 +285,11 @@ open_perf(pid_t target_tid, struct perf_pt_cerror *err) {
     // No skid.
     attr.precise_ip = 3;
 
+    // Wake up each time the AUX buffer becomes AUX_BUF_WAKE_PERCENT% full.
+    attr.watermark = 0;
+    attr.wakeup_events = 1;
+    attr.aux_watermark = (size_t) ((double) aux_bufsize * getpagesize()) * AUX_BUF_WAKE_PERCENT;
```

Comment:
I'm not sure I quite get what this line is doing.

Reply:
It's to do with the  Perf when to notifies us that the buffer is getting full.

From the `perf_event_open` manual:
```
       aux_watermark (since Linux 4.1)
              This specifies how much data is required to trigger a PERF_RECORD_AUX sample.
```

This means we wake up the poll loop when the buffer is `AUX_BUF_WAKE_PERCENT` full.

The AUX buffer size must be a power of two pages (our API lets the user configure the number of pages [hence `getpagesize()`], checking it's a power of 2 along the way).

Reply:
I sort of get that, but I don't get why we `* AUX_BUF_WAKE_PERCENT`. That desn't seem to match the description from the manual?

Reply:
` aux_bufsize * getpagesize()`  is the byte size of the buffer in total. `AUX_BUF_WAKE_PERCENT` is our own constant define as such:

```
#define AUX_BUF_WAKE_PERCENT 0.25
```

By multiplying by 0.25 we wake up the poll loop when the AUX buffer is roughly 25% full.

Perhaps that macro needs a better name?

[I think we will probably need to make this configurable in the future]

Reply:
A better name might not hurt. 25% also seems a really low value? I'm also not sure how keen I am on the floating point thing here. We might use this on machines where floating point calculations are really slow. Admittedly ` * 0.25` will presumably optimise to `>> 2` but maybe we should make that idea a bit more explicit? Perhaps only allow people to express how many bits to the right we'll shift?

Reply:
> We might use this on machines where floating point calculations are really slow.

I wouldn't worry about that. This code only executes on Intel machines new enough to have PT.

> 25% also seems a really low value?

It is. I've not yet decided a suitable default, but I do know that 75% with the default AUX buffer size causes underruns. How about 50%? [Of course a suitable value really depends how big your AUX buffer is.]

Reply:
50% seems a reasonable starting point.